src/server/command_registry.h

```diff
@@ -98,8 +101,30 @@ class CommandId {
     return step_key_;
   }
 
-  CommandId& SetHandler(Handler f) {
-    handler_ = std::move(f);
+  template <typename T> CommandId& AsyncHandler(T&& t) {
+    return *this;
+  }
+
+  template <typename F> CommandId& SetHandler(F&& f) {
+    using RET = std::invoke_result_t<F, CmdArgList, ConnectionContext*>;
+    if constexpr (std::is_void_v<RET>) {
+      handler_ = std::move(f);
+    } else {
+      static_assert(std::is_base_of_v<Responder, std::remove_pointer_t<RET>>);
+      opt_mask_ |= CO::ASYNC;
+      handler_ = [f = std::move(f)](CmdArgList args, ConnectionContext* cntx) {
+        // we manage responders inline, but in the future we can do whatever we want with them
+        // like sending over a channel, etc.
+        Responder* rsp = f(args, cntx);
+        bool done;
+        do {
+          done = rsp->Wait();
+          rsp->Respond(cntx);
+        } while (!done);
```

Comment:
The idea is the following: at the cost of one more call (nothing), we can now stream data, if it doesn't fit into one pass